### PR TITLE
Fix: DO-2870: Multiselect rerenders, migrate Multiselect to Floating UI

### PR DESCRIPTION
**Pull request by Roman-Kornev**
## Motivation and Context

MultiSelect component is rerendering all elements when scrolling even when it's off-screen or not even opened.

The performance hit is proportional to the number of elements inside the scrollbar. After around 800 elements the render time takes more than 5ms and it starts to become noticeable.

In some exteme cases this can increase the total page render time by **100x**.

On top of this, we're using an outdated version of Popper, which is no longer maintained. A more long-term solution would be to migrate over to Floating UI - its next evolution, which this PR also starts.

## Implementation Description

- Switched from `usePopper` to `useFloating`, maintaining all existing functionality.
- Fixed rerendering issue when scrolling. Seems to be caused by a feature similar to `autoUpdate`, when elements are hidden instead of unmounted when closed. (https://floating-ui.com/docs/autoUpdate). Fixed by conditionally adding `autoUpdate` only when open.
- Removed now redundant effect that was updating the floating element's position.

## Any new dependencies Introduced

Floating UI. The next evolution of Popper.
https://floating-ui.com/docs/migration

## How Has This Been Tested?

Locally in the storybook. All tests passing.
There is still one small issue of rerendering when the floating element is below the bottom edge of the screen that I haven't quite figured out, but it feels pretty minor compared to what it was before. Let me know if you have any ideas I might have missed (see gif).

## PR Checklist:

- [x] I have implemented all requirements? (see JIRA, project documentation).
- [x] I am not affecting someone else's work, If I am, they are included as a reviewer.
- [x] I have added relevant tests (unit, integration or regression).
- [x] I have added comments to all the bits that are hard to follow.
- [x] I have added/updated Documentation.
- [x] I have updated the appropriate changelog with a line for my changes.

## Screenshots (if appropriate):

BEFORE:
https://github.com/causalens/dara-ui/assets/119343173/7416556f-446a-4619-8621-16083ea0177c

AFTER:
https://github.com/causalens/dara-ui/assets/119343173/cce7e3f6-cdf2-4423-9fd3-0d127c28bdd6

### CODE REVIEW

File: packages/ui-components/src/multiselect/multiselect.tsx [hunk 1]
```diff
@@ -375,7 +400,6 @@ function MultiSelect({ maxWidth = '100%', maxRows = 3, ...props }: MultiSelectPr
                         <Input
                             {...getInputProps(getDropdownProps({ preventKeyAction: isOpen }))}
                             disabled={props.disabled}
-                            onFocus={openMenu}
```
Comment (krzysztof-causalens) [inline packages/ui-components/src/multiselect/multiselect.tsx:378]: Is this not required anymore?

**Review by krzysztof-causalens — APPROVED** [review]
Looks good, great find!

**Review by patricia-causalens — APPROVED** [review]
Hi Roman, nice work looks good!

What is the issue with the floating element being below the bottom edge of the screen? Is it rerendering too many times when outside view? Just curious

Also do we have a ticket to address this for normal select as well? I imagine it would suffer from the same problem

File: packages/ui-components/src/multiselect/multiselect.tsx (hunk 1 above)
Comment (Roman-Kornev) [reply packages/ui-components/src/multiselect/multiselect.tsx:378]: When I press Tab it automatically opens the menu, yes. Is there something else this was accounting for?

**Review by tamara-causalens — APPROVED** [review]
Nive, great work!

File: packages/ui-components/src/multiselect/multiselect.tsx (hunk 1 above)
Comment (krzysztof-causalens) [reply packages/ui-components/src/multiselect/multiselect.tsx:378]: No I think that's it, looks good then.

**Comment by Roman-Kornev** [comment]
> What is the issue with the floating element being below the bottom edge of the screen? Is it rerendering too many times when outside view? Just curious

Yeah, it's rerendering but only when opened and only at the bottom, not at the top. It's very bizarre.

> Also do we have a ticket to address this for normal select as well? I imagine it would suffer from the same problem

We have 6 other components using Popper right now, I'll go through them and apply similar fixes after this one.

**Pull request merged**